File: interview_cake/dynamic_programming_and_recursion/fibonacci.rb
def fib(n)

  # Compute the nth Fibonacci number.
  if n < 0
   raise ArgumentError, "There can't be a negative fibonacci number."
  elsif n <= 1
    return n
  end
  
  sum = 1
  last_number = 1
  (n - 2).times do 
    sum = sum + last_number
    last_number = sum - last_number
  end
  
  sum
end


















# Tests

def run_tests
  desc = 'zeroth fibonacci'
  actual = fib(0)
  expected = 0
  assert_equal(actual, expected, desc)

  desc = 'first fibonacci'
  actual = fib(1)
  expected = 1
  assert_equal(actual, expected, desc)

  desc = 'second fibonacci'
  actual = fib(2)
  expected = 1
  assert_equal(actual, expected, desc)

  desc = 'third fibonacci'
  actual = fib(3)
  expected = 2
  assert_equal(actual, expected, desc)

  desc = 'fifth fibonacci'
  actual = fib(5)
  expected = 5
  assert_equal(actual, expected, desc)

  desc = 'tenth fibonacci'
  actual = fib(10)
  expected = 55
  assert_equal(actual, expected, desc)

  desc = 'negative fibonacci'
  assert_raises(desc) { fib(-1) }
end

def assert_equal(a, b, desc)
  puts "#{desc} ... #{a == b ? 'PASS' : "FAIL: #{a.inspect} != #{b.inspect}"}"
end

def assert_raises(desc)
  yield
  puts "#{desc} ... FAIL"
rescue
  puts "#{desc} ... PASS"
end

run_tests

File: interview_cake/dynamic_programming_and_recursion/fibonacci.js
function fib(n) {

  // Compute the nth Fibonacci number
  if (n < 0) {
    throw "There can't be a negative fibonacci number."
  }
  
  if (n <= 1) {
    return n;
  }
  
  let sum = 1;
  let lastNumber = 1;
  for (let i = 0; i < n - 2; i++) {
    sum = sum + lastNumber;
    lastNumber = sum - lastNumber;
  }

  return sum;
}


















// Tests

let desc = 'zeroth fibonacci';
let actual = fib(0);
let expected = 0;
assertEqual(actual, expected, desc);

desc = 'first fibonacci';
actual = fib(1);
expected = 1;
assertEqual(actual, expected, desc);

desc = 'second fibonacci';
actual = fib(2);
expected = 1;
assertEqual(actual, expected, desc);

desc = 'third fibonacci';
actual = fib(3);
expected = 2;
assertEqual(actual, expected, desc);

desc = 'fifth fibonacci';
actual = fib(5);
expected = 5;
assertEqual(actual, expected, desc);

desc = 'tenth fibonacci';
actual = fib(10);
expected = 55;
assertEqual(actual, expected, desc);

desc = 'negative fibonacci';
const negativeFib = () => (fib(-1));
assertThrowsError(negativeFib, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`)
  }
}

function assertThrowsError(func, desc) {
  try {
    func();
    console.log(`${desc} ... FAIL`);
  } catch (e) {
    console.log(`${desc} ... PASS`);
  }
}
